### ISSUE TITLE
WASAPI backend: Libsoundio closes the program when no microphone is connected

### ISSUE THREAD

**Issue by zotta** [andrewrk/libsoundio#63]
In wsapi.c on line 694:

`if (FAILED(hr = IMMDeviceEnumerator_GetDefaultAudioEndpoint(siw->device_enumerator, eCapture,
                    eMultimedia, &rd.default_capture_device)))`

This command always fails if no microphone is connected at all. This leads to the program crashing on the next call of soundio_flush_events.

In other words: The library is basically unuseable on windows PCs without a microphone.

**Comment by andrewrk** [comment]
Nice find. Strange that it returns an error instead of a list with 0 items. And looks like you found 2 bugs in one with the crash. 

Let me get a patch going in a branch for you to try.

**Comment by andrewrk** [comment]
I'm going to try to get a VM here that doesn't have a mic to test on, but if you've already got this setup, could you determine which error code is returned from that API call in that case?

**Comment by zotta** [comment]
I am sorry for the late response, but i did have much time for my pet projects lately.
I changed it like this and it seems to work:
```
rd.default_render_device = NULL;
    if (FAILED(hr = IMMDeviceEnumerator_GetDefaultAudioEndpoint(siw->device_enumerator, eRender,
		eMultimedia, &rd.default_render_device)) && hr != HRESULT_FROM_WIN32(ERROR_NOT_FOUND))
    { [...]
    if (rd.default_render_device) { [...]
```

But your library turns out to be a multithreading nightmare. I basically would need a command that blocks/guarantees that all calls to the write callback have completed for my application to run properly.

**Comment by andrewrk** [comment]
> I basically would need a command that blocks/guarantees that all calls to the write callback have completed after pausing a stream, for my application to run properly.

I think the most portable thing to do here is destroy the output stream instead of pausing it. Then create it again to unpause.

**Comment by zotta** [comment]
The reason i tried to use pause is because destroying the stream in the middle of a callback causes all kinds of access violations in my application, so i tried pausing it instead.

**Comment by andrewrk** [comment]
Whoa, yes, you must not destroy the stream in the middle of a callback.

**Comment by andrewrk** [comment]
I believe @jacquesh fixed this with #79. Please re-open this issue if the problem occurs again.
